stdlib/public/Concurrency/TaskCancellation.swift

```diff
@@ -0,0 +1,81 @@
+////===----------------------------------------------------------------------===//
+////
+//// This source file is part of the Swift.org open source project
+////
+//// Copyright (c) 2020 Apple Inc. and the Swift project authors
+//// Licensed under Apache License v2.0 with Runtime Library Exception
+////
+//// See https://swift.org/LICENSE.txt for license information
+//// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+////
+////===----------------------------------------------------------------------===//
+
+import Swift
+import Dispatch
+@_implementationOnly import _SwiftConcurrencyShims
+
+// ==== Task Cancellation ------------------------------------------------------
+
+extension Task {
+  public struct CancellationError: Error {
+    // no extra information
+  }
+
+  /// Returns `true` if the task is cancelled.
+  public static func isCancelled() async -> Bool {
+    fatalError("\(#function) not implemented yet.")
+  }
+
+  /// Returns `true` if the task is not cancelled and should continue executing.
+  ///
+  /// This function is intended to be called in pair with `guard` statements:
+  ///
+  ///     func computePicture() async -> Picture {
+  ///         guard await Task.isNotCancelled() else {
+  ///             return .placeholder
+  ///         }
+  ///         return actuallyComputePicture()
+  ///     }
+  ///
+  /// or throwing a specific error rather than the generic cancellation error:
+  ///
+  ///     func computePicture() async throws -> Picture {
+  ///         guard await Task.isNotCancelled() else {
+  ///             throw PictureError.couldNotCompute
+  ///         }
+  ///         return actuallyComputePicture()
+  ///     }
+  ///
+  // TODO: decide if we want to keep this; it is useful in guard statements but
+  //       there isn't really precedent for isNot... APIs but guard is the primary use of these APIs
+  // TODO: is is typo-prone to offer isCancelled and isNotCancelled so if we
+  //       envision the use of guard statements to be common, it should be either
+  //       or "isCancelled" or the more guard friendly version, maybe we need a
+  //       new name for it?
+  public static func isNotCancelled() async -> Bool {
+    await !Task.isCancelled()
+  }
```

Comment:
This one is here for discussion, see the TODOs.

In theory one could imagine APIs which are inspecting the cancelled flag to provide a placeholder value, or log something and bail out / throw.

We do provide the simpler `await try checkCancellation()` which automatically throws the `Task.CancellationError` but we cannot prevent people from checking and doing something else on cancellation, so the question remains should we make it easy or hard to do and strongly push users to fo `checkCancellation()` instead.

WDYT @rjmccall @DougGregor ?